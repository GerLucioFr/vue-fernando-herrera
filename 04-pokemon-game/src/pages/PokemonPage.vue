<template>
  <div class="container">
    <h1>Who is this Pokemon?</h1>

    <div class="game-container" v-if="selectedPokemon && pokemonOptions.length">
      <PokemonPictureComponent
        :id="selectedPokemon.id"
        :showImage="chosenPokemon"
      />
      <PokemonOptionsComponent
        :pokemonOptions="pokemonOptions"
        :chosenPokemon="chosenPokemon"
        @selection="handleSelection"
      />

      <PokemonScoreComponent
        :currentScore="currentScore"
        :bestScore="bestScore"
      />

      <div class="fade-in" v-if="chosenPokemon && rightSelection">
        <h3>Excellent, is {{ selectedPokemon.name }}</h3>
        <button @click="resetTurn">Next</button>
      </div>

      <div class="fade-in" v-if="gameOver">
        <h2>Game Over</h2>
        <h4>Correct Pokemon was: {{ selectedPokemon.name }}</h4>
        <button @click="resetGame">Play again</button>
      </div>
    </div>

    <h3 v-else>Loading data...</h3>
  </div>
</template>

<script>
import PokemonPictureComponent from '@/components/pokemon/PokemonPictureComponent.vue';
import PokemonOptionsComponent from '@/components/pokemon/PokemonOptionsComponent.vue';
import PokemonScoreComponent from '@/components/pokemon/PokemonScoreComponent.vue';
import { getPokemonOptions } from '@/helpers/getPokemonOptions';
import { generateRandomNumber } from '@/helpers/number';
export default {
  name: 'PokemonPage',
  components: {
    PokemonOptionsComponent,
    PokemonPictureComponent,
    PokemonScoreComponent,
  },
  data() {
    return {
      pokemonOptions: [],
      selectedPokemon: null,
      chosenPokemon: false,
      rightSelection: false,
      gameOver: false,
      currentScore: 0,
      bestScore: 0,
    };
  },
  mounted() {
    this.getPokemons();
    this.getBestScore();
  },
  methods: {
    async getPokemons() {
      this.pokemonOptions = await getPokemonOptions(4);
      const selectedId = generateRandomNumber(
        0,
        this.pokemonOptions.length - 1
      );
      this.selectedPokemon = this.pokemonOptions[selectedId];
    },
    handleSelection(selection) {
      this.chosenPokemon = true;
      this.rightSelection = this.selectedPokemon.id === selection.id;

      if (this.rightSelection) {
        this.currentScore++;
        if (this.currentScore > this.bestScore) {
          this.setBesScore(this.currentScore);
        }
      } else {
        this.gameOver = true;
      }
    },
    resetGame() {
      this.currentScore = 0;
      this.gameOver = false;
      this.resetTurn();
    },
    resetTurn() {
      this.selectedPokemon = null;
      this.chosenPokemon = false;
      this.getPokemons();
    },
    getBestScore() {
      this.bestScore = Number(localStorage.getItem('bestScore'));
    },
    setBesScore(score) {
      this.bestScore = score;
      localStorage.setItem('bestScore', score);
    },
  },
};
</script>

<style>
button {
  background-color: rgb(54, 0, 104);
  color: white;
}

button:hover {
  background-color: rgb(33, 0, 65);
}
</style>
