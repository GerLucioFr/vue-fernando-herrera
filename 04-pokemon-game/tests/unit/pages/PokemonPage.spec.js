import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemonOptionsMock } from '../mocks/pokemons.mock';

describe('Pokemon Page Test', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  it('should to match with snapshot', () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonOptions: pokemonOptionsMock,
          selectedPokemon: pokemonOptionsMock[3],
          chosenPokemon: true,
          rightSelection: true,
          gameOver: false,
          currentScore: 1,
          bestScore: 0,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should to call getPokemons on mounted', () => {
    const getPokemonsSpy = jest.spyOn(PokemonPage.methods, 'getPokemons');
    shallowMount(PokemonPage);

    expect(getPokemonsSpy).toHaveBeenCalled();
  });

  it('should to exists child components', () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonOptions: pokemonOptionsMock,
          selectedPokemon: pokemonOptionsMock[3],
          chosenPokemon: true,
          rightSelection: true,
          gameOver: false,
          currentScore: 1,
          bestScore: 0,
        };
      },
    });

    expect(wrapper.find('.pokemon-container').exists()).toBeTruthy();
    expect(wrapper.find('.options-container').exists()).toBeTruthy();
    expect(wrapper.find('.score-container').exists()).toBeTruthy();
  });

  it('should to call handle answer method with right selection', async () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonOptions: pokemonOptionsMock,
          selectedPokemon: pokemonOptionsMock[3],
          chosenPokemon: false,
          rightSelection: false,
          gameOver: false,
          currentScore: 0,
          bestScore: 0,
        };
      },
    });
    const successMessage = `Excellent, is ${pokemonOptionsMock[3].name}`;
    const failMessage = `Correct Pokemon was: ${pokemonOptionsMock[3].name}`;

    await wrapper.vm.handleSelection(pokemonOptionsMock[3]);
    const subtitle = wrapper.find('h3');

    expect(wrapper.vm.chosenPokemon).toBeTruthy();
    expect(subtitle.text()).toBe(successMessage);

    await wrapper.vm.handleSelection(pokemonOptionsMock[2]);
    const subtitleError = wrapper.find('h4');
    expect(subtitleError.text()).toBe(failMessage);
  });
});
