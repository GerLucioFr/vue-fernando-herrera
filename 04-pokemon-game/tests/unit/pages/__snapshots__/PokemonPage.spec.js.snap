// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Page Test should to match with snapshot 1`] = `
<div class="container">
  <h1>Who is this Pokemon?</h1>
  <div class="game-container">
    <div class="pokemon-container"><img class="hidden-pokemon" src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/51.svg" alt="pokemon"><img class="fade-in" src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/51.svg" alt="pokemon"></div>
    <div class="options-container"><button disabled="">buneary</button><button disabled="">lanturn</button><button disabled="">minccino</button><button disabled="">dugtrio</button></div>
    <div class="score-container">
      <p>Current Score: 1</p>
      <p>Best Score: 0</p>
    </div>
    <div class="fade-in">
      <h3>Excellent, is dugtrio</h3><button>Next</button>
    </div>
    <!--v-if-->
  </div>
</div>
`;
